File: pipeline/geo.py
"""지리 계산 유틸. WGS84 위경도, Haversine 거리(m)."""
import math

_EARTH_R = 6371000.0  # 지구 반경(m)


def haversine(lat1: float, lng1: float, lat2: float, lng2: float) -> float:
    """두 위경도 사이 대권 거리(미터)."""
    if lat1 == lat2 and lng1 == lng2:
        return 0.0
    p1 = math.radians(lat1)
    p2 = math.radians(lat2)
    dphi = math.radians(lat2 - lat1)
    dlmb = math.radians(lng2 - lng1)
    a = math.sin(dphi / 2) ** 2 + math.cos(p1) * math.cos(p2) * math.sin(dlmb / 2) ** 2
    return 2 * _EARTH_R * math.asin(math.sqrt(a))

File: app/src/lib/geo.ts
// 지구 반지름(m). WGS84 위경도 두 점 사이 거리 계산(Haversine).
const EARTH_RADIUS_M = 6371000;

export interface LatLng {
  lat: number;
  lng: number;
}

/** 두 위경도 좌표 사이의 대권거리(미터). */
export function haversine(a: LatLng, b: LatLng): number {
  const toRad = (d: number) => (d * Math.PI) / 180;
  const dLat = toRad(b.lat - a.lat);
  const dLng = toRad(b.lng - a.lng);
  const lat1 = toRad(a.lat);
  const lat2 = toRad(b.lat);

  const h =
    Math.sin(dLat / 2) ** 2 +
    Math.cos(lat1) * Math.cos(lat2) * Math.sin(dLng / 2) ** 2;
  return 2 * EARTH_RADIUS_M * Math.asin(Math.min(1, Math.sqrt(h)));
}
